PXC-4173: PXC node stalls with parallel replication workers executing DDLs via async node

venkatesh-prasad-v opened the pull request:
https://perconadev.atlassian.net/browse/PXC-4173

    Problem
    =======
    PXC replica could enter into a deadlock with multi threaded replication when
    --replica-preserve-commit-order is enabled.
    
    The deadlock is more likely to occur on a replica with
    --replica-preserve-commit-order enabled when the source server's workload
    consists of concurrent executions of DMLs and DDLs. In such a scenario, the
    replica shall try to commit them in the same order as that of the source which
    will conflict with galera's ordering resulting in a deadlock.
    
    MySQL ordering: Commits are serialized as per their commit order on source server.
    Galera ordering: Commits are serialized in a FIFO manner.
    
    Scenario
    ========
    Let's assume we have the following two transactions in the relay log with same
    last_committed.
    
    T1: DML, seqno=10
    T2: DDL, seqno=11
    
    +-----------+--------------+-------------------------------------------------------+
    | Timestamp |    Thread    |                        Activity                       |
    +-----------+--------------+-------------------------------------------------------+
    |     t0    | Co-ordinator | Assigns T1 to Applier-1                               |
    +-----------+--------------+-------------------------------------------------------+
    |     t1    | Co-ordinator | Assigns T2 to Applier-2                               |
    +-----------+--------------+-------------------------------------------------------+
    |     t2    |   Applier-1  | DML starts with seqno = 10                            |
    +-----------+--------------+-------------------------------------------------------+
    |     t3    |   Applier-2  | DDL starts with seqno = 11                            |
    +-----------+--------------+-------------------------------------------------------+
    |     t4    |   Applier-2  | DDL calls TOI_begin(), acquires commit monitor        |
    +-----------+--------------+-------------------------------------------------------+
    |     t5    |   Applier-2  | DDL executes                                          |
    +-----------+--------------+-------------------------------------------------------+
    |     t5    |   Applier-1  | DML executes                                          |
    +-----------+--------------+-------------------------------------------------------+
    |     t6    |   Applier-2  | DDL reaches commit_order_manager, finds that          |
    |           |              | it needs to wait until Applier-1 is committed         |
    +-----------+--------------+-------------------------------------------------------+
    |     t7    |   Applier-1  | DML reaches commit, calls commit_order_enter_local()  |
    |           |              | and will wait for Applier-2 to release commit monitor |
    +-----------+--------------+-------------------------------------------------------+
    
    In the end it will result in
    Applier-1: DML executing Xid_log_event::commit waiting for DDL to release commit order
    Applier-2: DDL executing ALTER TABLE waiting for applier-1 to commit first
    
    Solution
    ========
    This commit introduces "Async Monitor" at server layer to ensure that the
    ordering in galera is done in the same order in the relay log.
    
    Async Monitor is similar to Commit Monitor used in galera, but differs in how
    seqnois are tracked within the monitor.
    
    The Async Monitor has the following methods.
    1. schedule() - Schedules the transactions in as per their order in the
                    the relay log.
    2. enter() - Ensures that threads are serialized as per their order.
    3. leave() - Ensures that threads are removed from the monitor in the serialized order.
    4. skip()  - Provides some flexibilty to the monitor to skip transactions. This
                 will be required for handling skipped and filtered transactions.
    
    Lifetime of the Async Monitor:
    Async Monitor is created on START REPLICA and is destroyed on STOP REPLICA, or
    whenever applier thread exits.

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/signed)](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1927) <br/>All committers have signed the CLA.

kamil-holubicki commented on sql/wsrep_async_monitor.cc:
```diff
@@ -0,0 +1,38 @@
+void Wsrep_async_monitor::leave(seqno_t seqno) {
+  // Wait for its turn before leaving
+  std::unique_lock<std::mutex> lock(m_mutex);
+  m_cond.wait(lock, [this, seqno]() { return seqno == m_last_left + 1; });
```
Is this cond.wait needed? I would add assert instead of this. The only valid situation is when the one who leaves monitor is 1 more than the previous one.

venkatesh-prasad-v replied:
Done

kamil-holubicki commented on sql/wsrep_mysqld.cc:
```diff
@@ -2870,6 +2870,57 @@ static void wsrep_RSU_end(THD *thd) {
   thd->disable_binlog_guard.reset();
 }
 
+bool thd_enter_apply_monitor(THD* thd) {
```
return value is not evaluated by the caller

venkatesh-prasad-v replied:
Changed the return type to `void` as `thd->killed` is the only condition that performs early return and it will be handled by the caller functions.

kamil-holubicki commented on sql/wsrep_mysqld.cc:
```diff
@@ -2870,6 +2870,57 @@ static void wsrep_RSU_end(THD *thd) {
 
+bool thd_enter_apply_monitor(THD* thd) {
+
+  if (thd->killed != THD::NOT_KILLED) {
```
shouldn't it be under if (thd->slave_thread)?

venkatesh-prasad-v replied:
Agree. Moved it to under `if (thd->slave_thread)`.

kamil-holubicki commented on sql/wsrep_mysqld.cc:
```diff
@@ -2870,6 +2870,57 @@ static void wsrep_RSU_end(THD *thd) {
+    return true;
+  }
+  if (thd->slave_thread) {
+    Slave_worker * sw = dynamic_cast<Slave_worker*>(thd->rli_slave);
```
Can we use static_cast here? If not, we should check for sw not being nullptr

venkatesh-prasad-v replied:
`Slave_worker` is a child class of `Relay_log_info` which is the type of `rli_slave`. So `dynamic_cast` is the better option here.

kamil-holubicki commented on sql/wsrep_async_monitor.cc:
```diff
@@ -0,0 +1,38 @@
+
+  // Wait for its turn before entering
+  m_cond.wait(lock, [this, seqno]() { return seqno == m_last_left + 1; });
+  m_last_entered = (seqno > m_last_entered) ? seqno : m_last_entered;
```
I don't understand this. Is it possible that seqno <= m_last_entered? I think the only possibility enforced by above condition.wait is that seqno grows monolithically.

venkatesh-prasad-v replied:
Hi! In an ideal case, it is not possible to have seqno < m_last_entered. I will change it just to `m_last_entered = seqno`

kamil-holubicki commented on sql/wsrep_async_monitor.cc:
```diff
@@ -0,0 +1,38 @@
+  // Wait for its turn before entering
+  m_cond.wait(lock, [this, seqno]() { return seqno == m_last_left + 1; });
+  m_last_entered = (seqno > m_last_entered) ? seqno : m_last_entered;
+  fprintf(stderr, "Entered the monitor with seqno: %llu\n", seqno);
```
Remember to remove fprintfs in the final version

venkatesh-prasad-v replied:
Sure. They are only added for QA. Once we get a green flag from the QA. I will remove these prints.

Commit:
PXC-4173: PXC node stalls with parallel replication workers executing DDLs via async node

https://perconadev.atlassian.net/browse/PXC-4173

Problem
=======
PXC replica could enter into a deadlock with multi threaded replication when
--replica-preserve-commit-order is enabled.

The deadlock is more likely to occur on a replica with
--replica-preserve-commit-order enabled when the source server's workload
consists of concurrent executions of DMLs and DDLs. In such a scenario, the
replica shall try to commit them in the same order as that of the source which
will conflict with galera's ordering resulting in a deadlock.

MySQL ordering: Commits are serialized as per their commit order on source server.
Galera ordering: Commits are serialized in a FIFO manner.

Scenario
========
Let's assume we have the following two transactions in the relay log with same
last_committed.

T1: DML, seqno=10
T2: DDL, seqno=11

+-----------+--------------+-------------------------------------------------------+
| Timestamp |    Thread    |                        Activity                       |
+-----------+--------------+-------------------------------------------------------+
|     t0    | Co-ordinator | Assigns T1 to Applier-1                               |
+-----------+--------------+-------------------------------------------------------+
|     t1    | Co-ordinator | Assigns T2 to Applier-2                               |
+-----------+--------------+-------------------------------------------------------+
|     t2    |   Applier-1  | DML starts with seqno = 10                            |
+-----------+--------------+-------------------------------------------------------+
|     t3    |   Applier-2  | DDL starts with seqno = 11                            |
+-----------+--------------+-------------------------------------------------------+
|     t4    |   Applier-2  | DDL calls TOI_begin(), acquires commit monitor        |
+-----------+--------------+-------------------------------------------------------+
|     t5    |   Applier-2  | DDL executes                                          |
+-----------+--------------+-------------------------------------------------------+
|     t5    |   Applier-1  | DML executes                                          |
+-----------+--------------+-------------------------------------------------------+
|     t6    |   Applier-2  | DDL reaches commit_order_manager, finds that          |
|           |              | it needs to wait until Applier-1 is committed         |
+-----------+--------------+-------------------------------------------------------+
|     t7    |   Applier-1  | DML reaches commit, calls commit_order_enter_local()  |
|           |              | and will wait for Applier-2 to release commit monitor |
+-----------+--------------+-------------------------------------------------------+

In the end it will result in
Applier-1: DML executing Xid_log_event::commit waiting for DDL to release commit order
Applier-2: DDL executing ALTER TABLE waiting for applier-1 to commit first

Solution
========
This commit introduces "Async Monitor" at server layer to ensure that the
ordering in galera is done in the same order in the relay log.

Async Monitor is similar to Commit Monitor used in galera, but differs in how
seqnois are tracked within the monitor.

The Async Monitor has the following methods.
1. schedule() - Schedules the transactions in as per their order in the
                the relay log.
2. enter() - Ensures that threads are serialized as per their order.
3. leave() - Ensures that threads are removed from the monitor in the serialized order.
4. skip()  - Provides some flexibilty to the monitor to skip transactions. This
             will be required for handling skipped and filtered transactions.

Lifetime of the Async Monitor:
Async Monitor is created on START REPLICA and is destroyed on STOP REPLICA, or
whenever applier thread exits.

This patch adds a new server variable `wsrep_use_async_monitor` for backward
compatibility

kamil-holubicki commented on sql/log_event.cc:
```diff
@@ -2704,6 +2704,14 @@ Slave_worker *Log_event::get_slave_worker(Relay_log_info *rli) {
+          if (wsrep_async_monitor) {
+            auto seqno = gtid_log_ev->sequence_number;
+            wsrep_async_monitor->schedule(seqno);
+            fprintf(stderr, "Scheduled the seqno: %llu in async monitor\n", seqno);
```
fprintf
multiple occurences

venkatesh-prasad-v replied:
I've added it for debugging purpose. I will remove all of them in the final version.

kamil-holubicki commented on sql/rpl_gtid_execution.cc:
```diff
@@ -37,6 +37,9 @@
 #include "sql/mysqld.h"  // connection_events_loop_aborted
 #include "sql/rpl_gtid.h"
 #include "sql/rpl_rli.h"    // Relay_log_info
+#ifdef WITH_WSREP
```
Can we keep WSREP-dependant includes after the standard ones, like we do in other cases?

venkatesh-prasad-v replied:
Done

kamil-holubicki commented on sql/rpl_rli.h:
```diff
@@ -63,6 +63,9 @@
 #include "sql/system_variables.h"
 #include "sql/table.h"
+#ifdef WITH_WSREP
+#include "sql/wsrep_async_monitor.h"
```
It looks like forward type declaration is enough, and the header can be included in *.cc file

venkatesh-prasad-v replied:
Done

kamil-holubicki commented on sql/wsrep_async_monitor.cc:
```diff
@@ -0,0 +1,94 @@
+  std::unique_lock<std::mutex> lock(m_mutex);
+
+  // Check if the sequence number matches the front of the queue.
+  // In a correctly functioning monitor this should not happen
```
In a correctly functioning monitor this should not happen ->
In a correctly functioning monitor this should always be true

venkatesh-prasad-v replied:
Done.

kamil-holubicki commented on sql/wsrep_async_monitor.cc:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  // Mark the seqno as skipped
+  skipped_seqnos.insert(seqno);
```
I can't see any place where seqnos are removed from `skipped_seqnos`

venkatesh-prasad-v replied:
Fixed it.

kamil-holubicki commented on sql/wsrep_mysqld.cc:
```diff
@@ -2858,6 +2858,61 @@ static void wsrep_RSU_end(THD *thd) {
+    if (wsrep_async_monitor) {
+      auto seqno = sw->sequence_number();
+      assert(seqno > 0);
+      // TODO: If requied, we must set current_mutex and current_cond here
```
TODO

venkatesh-prasad-v replied:
I would prefer having it as-is, as adding a new PFS state and using it with `enter_cond()` would make the thread's status visible in the SHOW PROCESSLIST. For example: we could make the thread to display `Waiting for async monitor` when it is waiting for its turn.

kamil-holubicki replied:
I'm just referring to leaving a TODO comment. Either implement it or remove the comment

kamil-holubicki commented on sql/wsrep_mysqld.cc:
```diff
@@ -2858,6 +2858,61 @@ static void wsrep_RSU_end(THD *thd) {
+      if (opt_replica_preserve_commit_order && opt_mts_replica_parallel_workers > 1) {
+        WSREP_ERROR("Async replica thread %u is not monitored by wsrep async monitor",
+                     thd->thread_id());
+        exit(1);
```
exit?
maybe unireg_abort(), or assert and handle it somehow?

venkatesh-prasad-v replied:
Good catch. Fixed.

kamil-holubicki commented on sql/wsrep_mysqld.cc:
```diff
@@ -2858,6 +2858,61 @@ static void wsrep_RSU_end(THD *thd) {
+      if (opt_replica_preserve_commit_order && opt_mts_replica_parallel_workers > 1) {
+        WSREP_ERROR("Async replica thread %u is not monitored by wsrep async monitor",
+                     thd->thread_id());
+        exit(1);
```
exit?

venkatesh-prasad-v replied:
Fixed

kamil-holubicki commented on sql/log_event.cc:
```diff
@@ -2704,6 +2704,14 @@ Slave_worker *Log_event::get_slave_worker(Relay_log_info *rli) {
+          Wsrep_async_monitor *wsrep_async_monitor {rli->get_wsrep_async_monitor()};
+          if (wsrep_async_monitor) {
+            auto seqno = gtid_log_ev->sequence_number;
+            wsrep_async_monitor->schedule(seqno);
```
Is my understandig correct, that in this point transactions appear in the order they were committed on the source node?
We create the queue of ordered transactions, and will use AsyncMonitor to enforce this order of commit in Galera, right?

venkatesh-prasad-v replied:
Yes. `schedule()`  is for updating the `scheduled_seqnos`  queue in the montior. This order defines the order in which the transactions will be entering the monitor.

kamil-holubicki commented on sql/wsrep_async_monitor.cc:
```diff
@@ -0,0 +1,94 @@
+// Method for main thread to add scheduled seqnos
+void Wsrep_async_monitor::schedule(seqno_t seqno) {
+  std::unique_lock<std::mutex> lock(m_mutex);
+  scheduled_seqnos.push(seqno);
```
seqons should be growing monolithically, right?
Can we add an assert like seqno = prev+1?
If not monolithically growing (why?), at least let's add an assert seqno > prev.
A good code comment would be appreciated

venkatesh-prasad-v replied:
In general, sequence numbers are generated monotonically. However that's only true within one binary log.  When the binlog gets rotated, the sequence number is reset and the first transaction in the newly created binlog will again start from seq no `1`. So, there is no need to add the assert `seqno = prev+1`.

kamil-holubicki commented on sql/wsrep_async_monitor.cc:
```diff
@@ -0,0 +1,94 @@
+
+  // Wait until this transaction is at the head of the scheduled queue
+  m_cond.wait(lock, [this, seqno] {
+    // Remove skipped transactions
```
When this can happen? Could you add a code comment? 
I suspect this is the case when A, B, C have been scheduled, then B skipped, but skip() removes only from the front of scheduled_seqno queue

venkatesh-prasad-v replied:
Yes. Imagine a scenario where scheduled seqnos is 1,2,3 and threads enter in an out of order manner.

- 3 enters the monitor first, since it is not in the front of the queue, it goes to cond_wait
- 2 enters the monitor, adds 2 to `skipped_seqnos` 
- 1 enters the monitor, since it is in the front, it acquires the monitor, does its job and leaves the monitor by removing itself from the `scheduled_seqnos` and signals 3.
- 3 wakes up, but it will see that 2 in the front of the queue. To proceed, it needs to remove all the skipped seqnos in the `scheduled_seqnos` queue.

kamil-holubicki replied:
Could you add a comment to the code explaining the purpose of this part?

venkatesh-prasad-v replied:
Done

kamil-holubicki commented:
The commit message says about ordering transactions basing on their seqnos i not way this = prev + 1. But the code does it in a different way (I can understand that logically it is still the same). It would be good to add more details about how the current implementation actually works.

venkatesh-prasad-v commented:
@kamil-holubicki Description is updated

venkatesh-prasad-v commented:
Jenkins: https://pxc.cd.percona.com/view/8.0%20parallel%20MTR/job/pxc-8.0-param-parallel-mtr/248/consoleFull

kamil-holubicki reviewed (approved):
LGTM
